### PR TITLE
Add basic Dreambox OS support

### PR DESCRIPTION
**Pull request by m0rphU**
The Dreambox OS has a few changes in the enigma2 API. This will apply the neccesary patches so that the OpenWebif will work there, too.

The main differences are:
- Changes in the enigma2 API for containers (`connect` instead of `append`)
- New Components for Network and Network
- `PluginDescriptor.WHERE_NETWORKCONFIG_READ` didn't seem to be working

A few minor things are still missing:
- [ ] Add .deb packaging script
- [ ] "Box Info": show all IPv6 information
- [ ] "Box Info": Detect real hard disk size
- [ ] Identify chipset, distro, driver date, kernel version and pictures of DreamOS boxes
- [ ] Transcoding
- [ ] Maybe add new DreamOS screenshot function instead of grab
- [ ] Maybe other instances of `e*Container()` that need to be changed?

Everything is tested with the latest DMM unstable on DM7080HD. It shouldn't break anything on other boxes

### CODE REVIEW

File: plugin/controllers/models/grab.py [hunk 1]
```diff
@@ -11,16 +11,21 @@
...
+DREAMOS = getInfo()['oever'] == "OE 2.2"
 
 class grabScreenshot(resource.Resource):
 	def __init__(self,session, path = ""):
 		resource.Resource.__init__(self)
 		self.session = session
 		self.container = eConsoleAppContainer()
-		self.container.appClosed.append(self.grabFinished)
-		# self.container.dataAvail.append(self.grabData)
+		if DREAMOS:
```
Comment (rdamas) [inline plugin/controllers/models/grab.py]: That looks ugly; do you want to make this kind of change for every eConsoleAppContainer() that will enter the code? Probably after you learned that it was added? Instead why not add a wrapper class that encapsulates the changes necessary for DREAMOS?

File: plugin/controllers/models/info.py [hunk 2]
```diff
@@ -257,60 +264,88 @@ def getInfo():
...
-			"mac": iNetwork.getAdapterAttribute(iface, "mac"),
-			"dhcp": iNetwork.getAdapterAttribute(iface, "dhcp"),
-			"ip": formatIp(iNetwork.getAdapterAttribute(iface, "ip")),
-			"mask": formatIp(iNetwork.getAdapterAttribute(iface, "netmask")),
-			"v4prefix": sum([bin(int(x)).count('1') for x in formatIp(iNetwork.getAdapterAttribute(iface, "netmask")).split('.')]),
-			"gw": formatIp(iNetwork.getAdapterAttribute(iface, "gateway")),
-			"ipv6": getAdapterIPv6(iface)['addr'],
-			"firstpublic": getAdapterIPv6(iface)['firstpublic']
-		})
+	if DREAMOS:
```
Comment (rdamas) [inline plugin/controllers/models/info.py]: The same holds here - create a wrapper that encapsulates the changes you want to add here.

**Comment by Schimmelreiter** [comment]
Another question:<br>
How will the patched version behave when (or if) some team decides to implement a more DreamOS compatible API, fully or partial?

**Comment by m0rphU** [comment]
Thank you for your comments. I agree that it's kind of hacky and ugly at the moment.

To be honest I didn't think of a wrapper yet, because I am not sure that this would be efficient. The changes are not that big, really. And I can't see how to do this without some if-else orgies :( Even in the wrapper this would be there.

Regarding the `eConsoleAppContainer`:
The change goes back to the DreamOS API change to `eSignal` as done in the DMM git with that commit: https://github.com/opendreambox/enigma2-plugins/commit/1f743037eaa4b631c73bfb64a33a9b3b1454f088
A lot of classes were affected. But as you see, most of the changes are only relevant for GUI (`eTimer`, `ePicload`, `ePythonMessagePump`).
I've rescanned the whole code in the OpenWebif for code that might need to be changed to `eSignal`. And besides the container for grab I found one in the ipkg code (not relevant for DreamOS). I also found some code in the stream controller that looks like it needs to be changed, but I have to verify that.
So what do you think here? Should I write a wrapper that will start an `eConsoleAppContainer` and if it is on DreamOS will return the connection object? Where should I put that wrapper?
But again, I am not sure that this is really necessary.

What I'd rather do is to change every place where it is necessary to something like that:
```
try:
  container.append()
except:
  conn = container.connect()
```
This has the benefit of not breaking anything, even if other teams will change to `eSignal` as well.

Now regarding the two other ugly changes (hdd and network):
I agree. This looks really strange. The changes are necessary because DMM removed/changed the relevant components because of changes under the hood (network is now maintained by ConnMan, harddisk is mainly taken care of by systemd).
But I already have an idea for wrappers there. I'll come back with a better version here :)


@Schimmelreiter: Well, there are a lot of things DreamOS does different (systemd, dpkg, the `eSignal` changes in e2-API, ...). So there always should be DreamOS specific code. But maybe that does not need to be generic but individually for each change.
Currently I have two different checks for DreamOS:
1. Rely on the `oever` info which is set to "OE 2.2" if dpkg is found
2. For network interfaces the converter uses a different name so I can try except the import

The first check will break as soon any image changes to dpkg as well but is not DreamOS or doesn't apply the other changes. And the second is also weird...
I'm not happy with both approaches because you check for things, that are not really relevant for the necessary changes

I've checked how other e2 plugins did that and most of them have some global variable which is set in plugin.py. I think we can do that with the info variable which is already global.
But how do we set this correctly? Most of the plugins check for a special DreamOS enigma module such as the mediadb. To me this is just as ugly as the two other methods I have right now.

I come to the follwoing conclusions:
- There shouldn't be just one single DreamOS variable like it is now. Each problem should be treated individually
- For the `eSignal` problem I don't see a generic solution. Maybe some try except orgies are best here?
- For specific changes to box information I will write specific wrappers
- I'd like to change the `PluginDescriptor` similar to the way it is done for VTI

Are you ok with that?

Thanks again,
m0rphU

P.S.: A git related question: Should I ammend my last commit? Or should I just push new commits? And will this stay in this pull request or should I start a new one if everything is cleared out?

**Comment by Schimmelreiter** [comment]
Well, one of the reasons why I asked:<br>
infos["dhcp"] = ip4.method

I was going to implement something similar in OpenATV (Actually I have a working set of .py with "ipmethod" and "ipv4method" instead of the silly "dhcp") but postponed the release because I first wanted to see if anyone else has done similar and keep compatibility to that implementation.
<br>
Obviously Dream shares the same logic as my approach by upgrading to a proper "method" info (Which could return manual, static, dhcp or disabled for legacy IPv4 and slaac, dhcp, static, manual, disabled for IP(v6)), so it is very likely that I will (at some time) create an implementation that is API compatible to iNetworkInfo.
<br>
So it would be unreliable to make a decision "DreamOS or not" based on single characteristics of DreamOS, as others might implement a subset of it.
<br>
systemd and dpkg/apt are also bad ways to detect DreamOS, as this is basically just two config values in some OpenEmbedded conf. Any distro can switch between systemd/sysvinit and opkg/dpkg any time.
<br>
<br>
The proper approach would be to create a new (wrapper) function to use for gathering this or that information and use that anywhere in the code.
The new function can then try both ways to gather the info (Just try the more advanced first), but then the two different ways are located in ONE place only.
<br>
<br>
About this pull request I would say we close it and you create a new one once you have something better.

**Comment by jbleyel** [comment]
Why do you don't use a global bool var for dreamOS and call the detection once?
For the grab.py and for the plugin.py and any other ..

But for info.py we shoud use a wrapper class to prevent this:

try:
 from Components.Network import iNetwork
except:
 from Components.Network import iNetworkInfo

**Pull request closed without merging**